### PR TITLE
Support published, private, and pending statuses

### LINKED ISSUE #12587: [Offline Posting] Support published, private, and pending statuses
📣 Part of #12227. Please target the `issue/12240-master-branch` branch. 
⚠️ Kinda blocked by the copy review of https://github.com/wordpress-mobile/WordPress-Android/pull/10542 but can still be worked on. 

At the time of this writing, all the auto-upload/auto-save/cancellation flow completed in `issue/12240-master-branch` only supports posts whose `status` (`BasePost.Status`) are:

- `.draft`
- `.publish`

## Goal

The goal is to support:

- `.publishPrivate`
- `.pending`
- `.scheduled`

These types are ignored and are out of scope:

- `.trash`
- `.deleted`

## Messages

Most of the work that will be done are adding messages. Please make sure that we are using the copy-reviewed messages as described in https://github.com/wordpress-mobile/WordPress-Android/pull/10542.

### LINKED ISSUE #12588: [Offline Posting] Publish buttons should not automatically publish scheduled posts
📣 Part of #12227. Please target the `issue/12240-master-branch` branch. 
⚠️ It's recommended to start this after #12587 is done.

This desired behavior is based on how Offline Posting for Android was implemented. 

## Scenario A

1. Go offline. 
2. Create a post using Gutenberg. 
3. Change the publishing schedule of the post in the Post Settings. 
4. Tap on _Publish_ in the Editor.
5. The Post List should show a message that the post will be scheduled when the device is online.
6. Press _Cancel_. The Post List should show ”Local changes“
7. Tap on the _Publish_ button in the Post List
8. Go online. The post should be scheduled and **not** immediately published. It should be visible in the _Scheduled_ list. 

## Scenario B

1. Go offline. 
2. Create a post using Gutenberg. 
3. Change the publishing schedule of the post in the Post Settings. 
4. Tap on _Publish_ in the Editor.
5. The Post List should show a message that the post will be scheduled when the device is online.
6. Press _Cancel_. The Post List should show ”Local changes“
7. Go online. 
8. Tap on the _Publish_ button in the Post List
9. The post should be scheduled and **not** immediately published. It should be visible in the _Scheduled_ list. 

---

FYI @yaelirub

### PR DESCRIPTION
**Pull request by leandroalonso**
Fixes #12587 and fixes #12588 

This will take some time to be reviewed. A few comments:

- Grab a giant cup of coffee before starting ☕️ 
- Don't limit yourself to test only the cases described here
- The strings are not final and will be updated in another task

# Private

## 1: Creating a private post and publishing it when online

1. While offline, create a private post and publish it
2. Check that the message "Private post will be published..." is appearing
3. Go back online, make sure the post was published and it is private

## 2: Changing an existent post to private

1. While online, create and publish a normal post
2. Go offline, edit the post and change it to private
3. Check if "Changes will be uploaded next time..." message is appearing
4. Go back online and make sure the post is now private

## 3: Multiple retries while offline

1. While offline, create a private post and publish it
2. Restart the app and check that the message shown is "Private post couldn't be published..." in yellow
3. Restart the app 2 more times
4. Check that the message shown is "Couldn't perform operation. Private post..." in red
5. Go back online, the post should not be auto upload

## 4: Cancel private post from being auto uploaded

1. While offline, create a private post and publish it
2. Tap cancel and check the snack bar message: "Changes will not be published"
3. Tap Publish, check that the post will be published as private later
4. Go back online, make sure the post is private

# Scheduled

## 1: Creating a scheduled post and publishing it when online

1. While offline, create a scheduled post and publish it
2. Check that the message "Post will be scheduled when your device is back online" is appearing
3. Go back online, make sure the post was scheduled

## 2: Changing an existent post to scheduled

1. While online, create and publish a normal post
2. Go offline, edit the post and change it to scheduled
3. Check if "Changes will be uploaded next time..." message is appearing
4. Go back online and make sure the post is now scheduled

## 3: Multiple retries while offline

1. While offline, create a scheduled post and publish it
2. Restart the app and check that the message shown is "Post couldn't be scheduled. We'll try again later" in yellow
3. Restart the app 2 more times
4. Check that the message shown is "Couldn't perform operation. Post not scheduled" in red
5. Go back online, the post should not be scheduled

## 4: Cancel scheduled post from being auto uploaded

1. While offline, create a private post and publish it
2. Tap cancel and check the snack bar message: "Changes will not be published"
3. Tap Publish, check that the post will be scheduled later
4. Go back online, make sure the post is scheduled

# Pending

For these tests you'd need a blog that you can only submit posts for review.

## 1: Creating a pending post and submitting it when online

1. While offline, create a post and submit it
2. Check that the message "Post will be submitted for review when your device is back online" is appearing
3. Go back online, make sure the post was submitted and is pending review

## 2: Multiple retries while offline

1. While offline, create a pending post and submit it
2. Restart the app and check that the message shown is "Post couldn't be submitted. We'll try again later" in yellow
3. Restart the app 2 more times
4. Check that the message shown is "Couldn't perform operation. Post not submitted" in red
5. Go back online, the post should not be submitted

## 3: Cancel pending post from being submitted

1. While offline, create a pending post and publish it
2. Tap cancel and check the snack bar message: "Changes will not be published"
3. Tap Publish, check that the post will be submitted later
4. Go back online, make sure the post is submitted

Update release notes:

- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.
- [x] If it's feasible, I have added unit tests.

### CODE REVIEW

File: WordPress/Classes/ViewRelated/Post/PostAutoUploadMessages.swift [hunk 1]
```diff
@@ -15,17 +15,72 @@ enum PostAutoUploadMessages {
...
+        case .publishPrivate:
+            return PostAutoUploadMessages.willAttemptToPublishPrivateLater
+        case .scheduled:
+            return PostAutoUploadMessages.willAttemptToScheduleLater
+        default:
+            return PostAutoUploadMessages.willAttemptToSubmitLater
+        }
     }
 
     static func willNotAttemptToAutoUpload(for postStatus: BasePost.Status?) -> String {
```
[the remark this answers is not in the document]
Comment (leandroalonso) [reply WordPress/Classes/ViewRelated/Post/PostAutoUploadMessages.swift]: Done in 7140a7a09f3dbea1c2084b89a71eaf8c66c3ded7

File: WordPress/Classes/ViewRelated/Post/PostAutoUploadMessages.swift [hunk 2]
```diff
@@ -15,17 +15,72 @@ enum PostAutoUploadMessages {
...
+    static let changesWillNotBePublished = NSLocalizedString("Changes will not be published",
+                                                        comment: "Title for notice displayed on canceling auto-upload published post")
+    static let changesWillNotBeSubmitted = NSLocalizedString("Changes will not be submitted",
+                                                         comment: "Title for notice displayed on canceling auto-upload pending post")
+    static let changesWillNotBeScheduled = NSLocalizedString("Changes will not be scheduled",
+                                                         comment: "Title for notice displayed on canceling auto-upload of a scheduled post")
+    static let changesWillNotBeSaved = NSLocalizedString("We won't save the latest changes to your draft.",
+                                                         comment: "Title for notice displayed on canceling auto-upload of a draft post")
 
     static func willAttemptToAutoUpload(for postStatus: BasePost.Status?) -> String {
```
[the remark this answers is not in the document]
Comment (leandroalonso) [reply WordPress/Classes/ViewRelated/Post/PostAutoUploadMessages.swift]: Done in 3fd58d4137092d4eb63d39192efb9b85032053c0

File: WordPress/WordPressTest/Services/PostAutoUploadInteractorTests.swift [hunk 3]
```diff
@@ -37,9 +37,9 @@ class PostCoordinatorUploadActionUseCaseTests: XCTestCase {
             // Posts with remote that have confirmation will be automatically uploaded
             createPost(.draft, hasRemote: true, confirmedAutoUpload: true): .upload,
             createPost(.publish, hasRemote: true, confirmedAutoUpload: true): .upload,
-            // Other statuses are currently ignored
-            createPost(.publishPrivate, confirmedAutoUpload: true): .nothing,
-            createPost(.scheduled, confirmedAutoUpload: true): .nothing,
+            createPost(.publishPrivate, confirmedAutoUpload: true): .upload,
+            createPost(.scheduled, confirmedAutoUpload: true): .upload,
```
[the remark this answers is not in the document]
Comment (leandroalonso) [reply WordPress/WordPressTest/Services/PostAutoUploadInteractorTests.swift]: Done in c16721593583a39a8cf398208818ba606f132d2e

File: WordPress/Classes/Services/PostService+MarkAsFailedAndDraftIfNeeded.swift [hunk 4]
```diff
@@ -14,19 +14,5 @@
         }
 
         post.remoteStatus = .failed
```
[the remark this answers is not in the document]
Comment (leandroalonso) [reply WordPress/Classes/Services/PostService+MarkAsFailedAndDraftIfNeeded.swift]: Good catch!

Done in e5fe285888674e9650893286a811b11cb2864454

Let me know what you think.

File: WordPress/Classes/ViewRelated/Post/AbstractPostListViewController.swift [hunk 5]
```diff
@@ -906,8 +906,14 @@ class AbstractPostListViewController: UIViewController,
...
+            if apost.status == .draft {
+                apost.status = .publish
+            }
+
+            if apost.status != .scheduled {
+                apost.date_created_gmt = Date()
+            }
+
             apost.shouldAttemptAutoUpload = true
             self.uploadPost(apost)
```
[the remark this answers is not in the document]
Comment (leandroalonso) [reply WordPress/Classes/ViewRelated/Post/AbstractPostListViewController.swift]: Done.

After looking at this refactor, it seems to me that `PostCoordinator` and `PublishPostInteractor` are kinda the same things with different names. What do you think?

File: WordPress/Classes/ViewRelated/Post/PostEditor+Publish.swift [hunk 6]
```diff
@@ -84,6 +84,8 @@ extension PostEditor where Self: UIViewController {
                 if self.post.status != .publishPrivate {
                     self.post.status = .publish
                 }
+            } else if action == .submitForReview {
+                self.post.status = .pending
```
[the remark this answers is not in the document]
Comment (leandroalonso) [reply WordPress/Classes/ViewRelated/Post/PostEditor+Publish.swift]: Done in d8624ec6e554e30c02e0047c63a6304d46567ef8

File: WordPress/WordPressTest/Services/PostCoordinatorFailedPostsFetcherTests.swift [hunk 7]
```diff
@@ -32,12 +32,14 @@ class PostCoordinatorFailedPostsFetcherTests: XCTestCase {
             createPost(status: .publish),
             createPost(status: .draft, hasRemote: true),
             createPost(status: .publish, hasRemote: true),
-        ]
-        let unexpectedPosts = [
             createPost(status: .publishPrivate),
             createPost(status: .publishPrivate, hasRemote: true),
             createPost(status: .scheduled),
             createPost(status: .scheduled, hasRemote: true),
+            createPost(status: .pending),
```
[the remark this answers is not in the document]
Comment (leandroalonso) [reply WordPress/WordPressTest/Services/PostCoordinatorFailedPostsFetcherTests.swift]: Done in 515ed05d2daf7b6ef085076004f3a8057c052d22

**Comment by leandroalonso** [comment]
@shiki Available for another pair of eyes. 👀 

About the out of scope bug you mentioned, I created an issue for it: https://github.com/wordpress-mobile/WordPress-iOS/issues/12630

File: WordPress/Classes/ViewRelated/Post/AbstractPostListViewController.swift (hunk 5 above)
[the remark this answers is not in the document]
Comment (shiki) [reply WordPress/Classes/ViewRelated/Post/AbstractPostListViewController.swift]: You're probably right. Should we move them there then? As long as it's tested, I'm okay with it.

**Comment by shiki** [comment]
@leandroalonso Changes look good. I replied to your comments. There are 2 outstanding discussions left: 

- https://github.com/wordpress-mobile/WordPress-iOS/pull/12613#commitcomment-35391692
- https://github.com/wordpress-mobile/WordPress-iOS/pull/12613#discussion_r331658706

File: WordPress/Classes/ViewRelated/Post/AbstractPostListViewController.swift (hunk 5 above)
[the remark this answers is not in the document]
Comment (leandroalonso) [reply WordPress/Classes/ViewRelated/Post/AbstractPostListViewController.swift]: Done.

**Comment by leandroalonso** [comment]
@shiki again, ready for another look. :)

**Pull request merged**